### PR TITLE
OSDOCS-476 Document Azure registry storage

### PR DESCRIPTION
**Pull request by bmcelvee**
Stub for documenting https://jira.coreos.com/browse/OSDOCS-476.

### CODE REVIEW

**Comment by bmcelvee** [comment]
Hi @dmage, I opened this PR to start documenting the Azure registry storage set-up for https://jira.coreos.com/browse/DEVEXP-147. I looked though the Operator code and didn't see parameters for Azure beyond `ImageRegistryConfigStorageAzure`.

Does the user need to do anything beyond providing cloud credentials to set up the Azure registry storage? 

Thanks!

**Comment by dmage** [comment]
@bmcelvee we have two supported use case: IPI and UPI.

On IPI the registry will be configured automatically, you don't need to provide anything.

On UPI the user should provide the secret `image-registry-private-configuration-user` in the `openshift-image-registry` namespace with the key `REGISTRY_STORAGE_AZURE_ACCOUNTKEY`. And they also will need to create an Azure storage container and provide both parameters (accountName and container) for ImageRegistryConfigStorageAzure on the config image registry resource (spec.storage.azure).

**Comment by bmcelvee** [comment]
Thanks, @dmage! I updated a note in https://github.com/openshift/openshift-docs/pull/16248 stating that on IPI the registry will be configured automatically. Once that PR merges, I'll be able update the procedure for adding the Secret to the appropriate module and include it in the configuring Azure registry storage assembly.

File: modules/registry-operator-config-resources-secret.adoc [hunk 1]
```diff
@@ -45,12 +49,20 @@ $ oc create secret generic image-registry-private-configuration-user --from-lite
...
 ----
 $ oc create secret generic image-registry-private-configuration-user --from-file=REGISTRY_STORAGE_GCS_KEYFILE=<path_to_keyfile> --namespace openshift-image-registry
 ----
 
+*For Azure:*
+
+* Create an {product-title} Secret that contains the required key.
++
+----
+$ oc create secret generic image-registry-private-configuration-user --from-file=REGISTRY_STORAGE_AZURE_ACCOUNTKEY=<path_to_accountkey> --namespace openshift-image-registry
```
Comment (dmage) [inline modules/registry-operator-config-resources-secret.adoc]: An Azure access key is a string that you can copy from the Azure web console, so I think it's better to use `--from-literal=REGISTRY_STORAGE_AZURE_ACCOUNTKEY=<accountkey>`

File: modules/registry-configuring-storage-azure-user-infra.adoc [hunk 2]
```diff
@@ -0,0 +1,24 @@
+// Module included in the following assemblies:
+//
+//* registry/configuring-registry-storage-azure.adoc
+
+[id="registry-configuring-storage-azure-user-infra_{context}"]
+= Configuring registry storage for Azure
+
+During installation, your cloud credentials are sufficient to create Azure Blob
+Storage and the Registry Operator will automatically configure storage. Azure
+registry storage uses the struct, `ImageRegistryConfigStorageAzure`.
```
Comment (dmage) [inline modules/registry-configuring-storage-azure-user-infra.adoc]: Is there any place where customers can see `ImageRegistryConfigStorageAzure`?

File: _topic_map.yml [hunk 3]
```diff
@@ -349,18 +349,25 @@ Topics:
...
+  - Name: Configuring registry storage for AWS user-provisioned infrastructure
+    File: configuring-registry-storage-aws-user-infrastructure
+    Distros: openshift-enterprise,openshift-origin
+  - Name: Configuring registry storage for GCP user-provisioned infrastructure
+    File: configuring-registry-storage-gcp-user-infrastructure
+    Distros: openshift-enterprise,openshift-origin
+  - Name: Configuring registry storage for Azure user-provisioned infrastructure
+    File: configuring-registry-storage-azure-user-infrastructure
+    Distros: openshift-enterprise,openshift-origin
+  - Name: Configuring registry storage for baremetal
```
Comment (kalexand-rh) [inline _topic_map.yml]: bare metal

File: modules/registry-configuring-storage-azure-user-infra.adoc [hunk 4]
```diff
@@ -0,0 +1,24 @@
+// Module included in the following assemblies:
+//
+//* registry/configuring-registry-storage-azure.adoc
+
+[id="registry-configuring-storage-azure-user-infra_{context}"]
+= Configuring registry storage for Azure
+
+During installation, your cloud credentials are sufficient to create Azure Blob
+Storage and the Registry Operator will automatically configure storage. Azure
```
Comment (kalexand-rh) [inline modules/registry-configuring-storage-azure-user-infra.adoc]: s/Storage/Storage,
s/will automatically configure/automatically configures

File: modules/registry-configuring-storage-azure-user-infra.adoc (hunk 2 above)
Comment (kalexand-rh) [reply modules/registry-configuring-storage-azure-user-infra.adoc]: s/struct, `ImageRegistryConfigStorageAzure`./`ImageRegistryConfigStorageAzure` struct.

File: modules/registry-configuring-storage-azure-user-infra.adoc [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+registry storage uses the struct, `ImageRegistryConfigStorageAzure`.
+
+.Prerequisites
+
+* A cluster on Azure with user-provisioned infrastructure.
```
Comment (kalexand-rh) [inline modules/registry-configuring-storage-azure-user-infra.adoc]: Azure UPI isn't in 4.2.

File: modules/registry-configuring-storage-azure-user-infra.adoc [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+* To configure registry storage for Azure, you need to provide Registry Operator
```
Comment (kalexand-rh) [inline modules/registry-configuring-storage-azure-user-infra.adoc]: ~you need to~

File: modules/registry-configuring-storage-azure-user-infra.adoc [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+cloud credentials.
+* For Azure storage the Secret is expected to contain one key:
```
Comment (kalexand-rh) [inline modules/registry-configuring-storage-azure-user-infra.adoc]: Maybe 

```
* The Secret for Azure storage contains only the `REGISTRY_STORAGE_AZURE_ACCOUNTKEY` key.
```

File: modules/registry-configuring-storage-azure-user-infra.adoc [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+** `REGISTRY_STORAGE_AZURE_ACCOUNTKEY`
+
+.Procedure
+
+ * Create an Azure storage container and provide the `accountName` and
```
Comment (kalexand-rh) [inline modules/registry-configuring-storage-azure-user-infra.adoc]: Do you have a link in the Azure docs that describes this process? If so, I'd add it.

**Review by kalexand-rh — COMMENTED** [review]
Just a few picks.

**Comment by bmcelvee** [comment]
Thank you! I will address feedback soon. At this point it looks like UPI installation is not available in 4.2.  
At this point I think I have two options for this 4.2:

1. Comment out the UPI content until it GAs
2. Put a hold on this PR until it GAs

File: modules/registry-configuring-storage-azure-user-infra.adoc [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+ `container` parameters for `ImageRegistryConfigStorageAzure` on the
+ configuration image registry resource `spec.storage.azure` file.
```
Comment (xiuwang) [inline modules/registry-configuring-storage-azure-user-infra.adoc]: Better points 'the
 configuration image registry resource' as 'configs.imageregistry.operator.openshift.io' explicitly

**Comment by xiuwang** [comment]
@bmcelvee There is no description for azure storage parameters like [gcs ](https://docs.openshift.com/container-platform/4.2/registry/configuring-registry-storage-gcp-user-infrastructure.html#registry-operator-configuration-resource-overview-gcp-gcs_configuring-registry-storage-gcp-user-infrastructure)and aws do,

**Comment by vikram-redhat** [comment]
@bmcelvee can this be marked as done for QE to do a final review?

**Comment by bmcelvee** [comment]
@vikram-redhat Yep! It's ready for final QE review. The build failure is not related to this PR, so I'll rebase when the issue is resolved. 

@xiuwang please take a look when you get a chance, thanks!

File: modules/registry-configuring-storage-azure-user-infra.adoc [hunk 10]
```diff
@@ -0,0 +1,24 @@
+// Module included in the following assemblies:
+//
+//* registry/configuring-registry-storage-azure.adoc
+
+[id="registry-configuring-storage-azure-user-infra_{context}"]
+= Configuring registry storage for Azure
+
+During installation, your cloud credentials are sufficient to create Azure Blob
+Storage, and the Registry Operator automatically configures storage. Azure
+registry storage uses the `ImageRegistryConfigStorageAzure` struct.
```
Comment (xiuwang) [inline modules/registry-configuring-storage-azure-user-infra.adoc]: `ImageRegistryConfigStorageAzure` struct is in operator code, the customers have no way to see this in `configs.imageregistry.operator.openshift.io/cluster`, do we need remove this sentence？

File: modules/registry-configuring-storage-azure-user-infra.adoc (hunk 5 above)
Comment (xiuwang) [reply modules/registry-configuring-storage-azure-user-infra.adoc]: Azure UPI is not ready for test in 4.2

File: modules/registry-operator-config-resources-secret-azure.adoc [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+In addition to the `configs.imageregistry.operator.openshift.io` and ConfigMap
+resources, configuration is provided to the Operator by a separate Secret
+resource located within the `openshift-image-registry` namespace.
+
+The `image-registry-private-configuration-user` Secret provides
+credentials needed for storage access and management. It overrides the default
+credentials used by the Operator, if default credentials were found.
+
+For Azure registry storage the secret is expected to contain one key whose value is the
```
Comment (xiuwang) [inline modules/registry-operator-config-resources-secret-azure.adoc]: s/the secret/the Secret
Keep consistent with gcp|aws file

**Comment by xiuwang** [comment]
@bmcelvee Just adding comments in two minior places, you can judge them.
Other lgtm.

File: modules/registry-configuring-storage-azure-user-infra.adoc (hunk 10 above)
Comment (bmcelvee) [reply modules/registry-configuring-storage-azure-user-infra.adoc]: Thank you for the clarification! I will remove the sentence.

**Comment by bmcelvee** [comment]
Thanks @xiuwang! I made the changes and am going ahead with the merge and cherrypick.

**Comment by vikram-redhat** [comment]
@bmcelvee not sure what happened here, but looks like Travis failed us big time! 

The build hadn't passed: https://travis-ci.com/openshift/openshift-docs/builds/131319073#L266, yet, I can see that here that it said that it passed. I don't know why the build didn't fail it. *grumble*

In any case, I have pushed an update to fix this via: https://github.com/openshift/openshift-docs/pull/17205

**Comment by bmcelvee** [comment]
@vikram-redhat thank you so much! I thought it had passed. :( Travis was acting strangely yesterday in general.

**Pull request merged**